### PR TITLE
 Implement Stream.WriteTo with custom offset

### PR DESCRIPTION
**Pull request by Megaflan**
### Description
Now there's a `Stream.WriteSegmentTo` function so you can provide a custom offset to it and it will write in the determined position, it's just a change in the use of `Seek` in the `WriteTo` function.
`WriteTo` also changed to work with `WriteSegmentTo(0, stream);`
As @pleonex suggested, there's also an overload of this function with a custom length, the function will define the size of the buffer with the provided length instead of reading the length of the stream.

### Example
`WriteSegmentTo(long start, DataStream stream)`
`WriteSegmentTo(long start, long length, DataStream stream)`

Linked issue:
#73

### CODE REVIEW

File: src/Yarhl.UnitTests/IO/DataStreamTests.cs [hunk 1]
```diff
@@ -1387,6 +1387,111 @@ public void WriteToAfterDispose()
...
+        public void WriteSegmentToFileCreatesParentFolder()
+        {
+            string tempFile = Path.Combine(
+                Path.GetTempPath(),
+                Path.GetRandomFileName(),
+                Path.GetRandomFileName());
+
+            DataStream stream = new DataStream();
+            stream.WriteByte(0xCA);
+            stream.WriteByte(0xFE);
```
Comment (pleonex) [inline src/Yarhl.UnitTests/IO/DataStreamTests.cs:1501]: You need to write more data here so we can ensure we only write to the file a part of it.

File: src/Yarhl.UnitTests/IO/DataStreamTests.cs [hunk 2]
```diff
@@ -1387,6 +1387,111 @@ public void WriteToAfterDispose()
...
+            stream.WriteSegmentTo(0, 2, tempFile);
```
Comment (pleonex) [inline src/Yarhl.UnitTests/IO/DataStreamTests.cs:1506]: Can you use offset = 1 so we test also that we are changing the position?

File: src/Yarhl.UnitTests/IO/DataStreamTests.cs [hunk 3]
```diff
@@ -1387,6 +1387,111 @@ public void WriteToAfterDispose()
...
+        [Test]
+        public void WriteSegmentToVariableLength()
+        {
+            DataStream stream1 = new DataStream();
+            stream1.WriteByte(0xCA);
+            stream1.WriteByte(0xFE);
+            stream1.WriteByte(0x00);
+            stream1.WriteByte(0xFF);
+            DataStream stream2 = new DataStream();
+            stream1.WriteSegmentTo(0, 2, stream2);
```
Comment (pleonex) [inline src/Yarhl.UnitTests/IO/DataStreamTests.cs:1399]: Can you use offset 1 so we test that we are changing the position?

File: src/Yarhl.UnitTests/IO/DataStreamTests.cs [hunk 4]
```diff
@@ -1387,6 +1387,111 @@ public void WriteToAfterDispose()
             Assert.Throws<ObjectDisposedException>(() => stream1.WriteTo("/ex"));
         }
 
+        [Test]
+        public void WriteSegmentToVariableLength()
```
Comment (pleonex) [inline src/Yarhl.UnitTests/IO/DataStreamTests.cs:1391]: Can you add another test to validate the overload `WriteSegmentTo(offset, stream)`, where offset is different to 0? This way we guarantee we are calculating the correct length always.

File: src/Yarhl/IO/DataStream.cs [hunk 5]
```diff
@@ -497,19 +492,102 @@ public void WriteTo(DataStream stream)
...
 
             long currPos = Position;
-            Seek(0, SeekMode.Start);
+            Seek(start, SeekMode.Start);
 
             const int BufferSize = 70 * 1024;
             byte[] buffer = new byte[Length > BufferSize ? BufferSize : Length];
 
-            while (!EndOfStream) {
+            while (!EndOfStream)
```
Comment (pleonex) [inline src/Yarhl/IO/DataStream.cs]: The code convention is to put the braces in the same line :)

File: src/Yarhl/IO/DataStream.cs [hunk 6]
```diff
@@ -497,19 +492,102 @@ public void WriteTo(DataStream stream)
...
             Seek(currPos, SeekMode.Start);
         }
 
+        /// <summary>
+        /// Writes a defined length stream into another DataStream starting in a defined position.
+        /// </summary>
+        /// <param name="start">Defined starting position.</param>
+        /// <param name="length">Defined length to be written.</param>
+        /// <param name="stream">Output DataStream.</param>
+        public void WriteSegmentTo(long start, int length, DataStream stream)
```
Comment (pleonex) [inline src/Yarhl/IO/DataStream.cs]: Length should be a `long` value to be consistent with the type of the property `Length`.

File: src/Yarhl/IO/DataStream.cs [hunk 7]
```diff
@@ -497,19 +492,102 @@ public void WriteTo(DataStream stream)
...
+
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
+            if (stream.Disposed)
+                throw new ObjectDisposedException(nameof(stream));
+
+            long currPos = Position;
+            Seek(start, SeekMode.Start);
+
+            byte[] buffer = new byte[length];
```
Comment (pleonex) [inline src/Yarhl/IO/DataStream.cs]: This may cause memory issues if length is too big. For instance imagine calling this method for a big CVM file to remove the header. We should do the same as WriteSegmentTo without length does. In fact, what about if we adapt that logic here and that method calls this one with length = `Length`?

**Review by pleonex — CHANGES_REQUESTED** [review]
It looks very good, thanks!
Just a few changes to improve the performance and to test tricky corner cases.

File: src/Yarhl/IO/DataStream.cs (hunk 5 above)
Comment (Megaflan) [reply src/Yarhl/IO/DataStream.cs]: Done

File: src/Yarhl/IO/DataStream.cs (hunk 6 above)
Comment (Megaflan) [reply src/Yarhl/IO/DataStream.cs]: Done

**Review by pleonex — APPROVED** [review]
Everything is perfect now. Thanks a lot!
